### PR TITLE
stm32h7: Updated pwr and rcc configs to support devices with SMPS.

### PR DESCRIPTION
**Pull request by vielster**
Worked in nuances for differences between versions of STM32H7 devices, such as
handling of ODEN, explicit SCUEN bit, and different VOS mappings. This has
been validated on the STM32H7A3 and STM32H743 MCUs.

### CODE REVIEW

File: include/libopencm3/stm32/dbgmcu.h [hunk 1]
```diff
@@ -37,6 +37,23 @@
...
+	DBGMCU_IDCODE_DEV_ID_STM32H750 = 0x450,
+	DBGMCU_IDCODE_DEV_ID_STM32H753 = 0x450,
+	DBGMCU_IDCODE_DEV_ID_STM32H755 = 0x450,
+	DBGMCU_IDCODE_DEV_ID_STM32H757 = 0x450,
+	DBGMCU_IDCODE_DEV_ID_STM32H7A3 = 0x480,
+	DBGMCU_IDCODE_DEV_ID_STM32H7B3 = 0x480,
+	DBGMCU_IDCODE_DEV_ID_STM32H7B0 = 0x480,
+	DBGMCU_IDCODE_DEV_ID_STM32H72X = 0x483,
+	DBGMCU_IDCODE_DEV_ID_STM32H73X = 0x483,
+};
```
Comment (karlp) [inline include/libopencm3/stm32/dbgmcu.h:55]: I don't really feel this is appropriate honestly, at least, certainly not dropped into the h7 pwr commit, and adding a limited subset of h7s to the shared file. I'm not _really_ sure it should ever be defined honestly, but definitely not here in this commit

Comment (vielster) [reply include/libopencm3/stm32/dbgmcu.h:55]: Would you be okay with #defines inside of pwr.c that hold the two constants I care about...something like

```c++
#define DBGMCU_IDCODE_DEV_ID_STM32H74X_5X		0x450
#define DBGMCU_IDCODE_DEV_ID_STM32H7A3_B3_B0		0x480
```

File: include/libopencm3/stm32/h7/memorymap.h [hunk 2]
```diff
@@ -154,4 +154,7 @@
 #define TIM3_BASE           0x40000400U
 #define TIM2_BASE           0x40000000U
 
+/* APB-D Peripherals */
+#define DBGMCU_BASE         0x5C001000
```
Comment (karlp) [inline include/libopencm3/stm32/h7/memorymap.h:158]: again, not part of this commit, but totally should come in of course.  however, I think the comment is wrong?

`They are accessible to the debugger via the APB-D bus at base address 0xE00E1000. They are also
accessible by the processor core at base address 0x5C001000.` 

 (just drop the comment maybe?)

File: include/libopencm3/stm32/h7/pwr.h [hunk 3]
```diff
@@ -57,39 +61,133 @@ LGPL License Terms @ref lgpl_license
...
 #define PWR_D3CR_VOS_SCALE_1      (0x3)
-#define PWR_D3CR_VOS_MASK         (0x03)
+/** SRDCR Register Bits */
+#define PWR_SRDCR_VOSRDY          BIT13
+#define PWR_SRDCR_VOS_SHIFT       14
+#define PWR_SRDCR_VOS_MASK        (0x03)
+#define PWR_SRDCR_VOS_SCALE_3     (0x0)
+#define PWR_SRDCR_VOS_SCALE_2     (0x1)
+#define PWR_SRDCR_VOS_SCALE_1     (0x2)
+#define PWR_SRDCR_VOS_SCALE_0     (0x3)
```
Comment (karlp) [inline include/libopencm3/stm32/h7/pwr.h:179]: oh that's gross, boo ST! :)

File: lib/stm32/h7/pwr.c [hunk 4]
```diff
@@ -30,15 +30,79 @@
...
+		case PWR_SYS_EXT_SMPS_LDO:
+		case PWR_SYS_EXT_SMPS_LDO_BYP:
+			reg_level = smps_level << PWR_CR3_SMPSLEVEL_SHIFT;
+			pwr_set_mode_smps_ldo(false, reg_level, mode == PWR_SYS_EXT_SMPS_LDO);
+			break;
+		case PWR_SYS_BYPASS:
+			pwr_set_mode_bypass();
+			break;
+	}
+	// Wait for power supply status to state ready.
```
Comment (karlp) [inline lib/stm32/h7/pwr.c:105]: no c++ comments please

File: lib/stm32/h7/pwr.c [hunk 5]
```diff
@@ -50,16 +114,49 @@ void pwr_set_svos_scale(enum pwr_svos_scale scale)
...
 
-	/* Per the manual, VOS0 is implemented as VOS1 + ODEN. Handle this case. */
-	if (scale == PWR_VOS_SCALE_0) {
-		PWR_D3CR = d3cr_masked | PWR_VOS_SCALE_1;
-		SYSCFG_PWRCR |= SYSCFG_PWRCR_ODEN;
+	/* "SmartRun Domain" devices (presently only know of A3/B3/B0) have different mapping.
+	 * Note: DBGMCU_IDCODE_DEV_ID_STM32H7A3 covers all three of these models.
+	 */
+	uint32_t devid = DBGMCU_IDCODE & DBGMCU_IDCODE_DEV_ID_MASK;
+	if (devid == DBGMCU_IDCODE_DEV_ID_STM32H7A3) {
```
Comment (karlp) [inline lib/stm32/h7/pwr.c:135]: I see why you're doing this, but this sort of stuff is internal to this file, you don't want to go adding all those id codes to the global file.

File: include/libopencm3/stm32/dbgmcu.h (hunk 1 above)
Comment (karlp) [reply include/libopencm3/stm32/dbgmcu.h:55]: yep, was going to say that now that I'd gotten to the bottom, just not fast enough :)

**Comment by karlp** [comment]
this looks just about exactly how it shoudl get done really :) @donotnoot does that work for you?  It looks good to me, on the face of it, if vielster cleans up the dbgmcu stuff

File: include/libopencm3/stm32/h7/pwr.h (hunk 3 above)
Comment (vielster) [reply include/libopencm3/stm32/h7/pwr.h:179]: Tell me about it!  Also, fun debugging this stuff because the LDO/SMPS/etc settings (not including VOS) can only be written once per power cycle :-)

File: include/libopencm3/stm32/h7/memorymap.h (hunk 2 above)
Comment (vielster) [reply include/libopencm3/stm32/h7/memorymap.h]: Made comment generic...there are a few other Debug/Trace periphs in the same range that may get added later...I feel weird having no separator at all since it's not an APB1 peripheral :-)

File: lib/stm32/h7/pwr.c [hunk 6]
```diff
@@ -30,15 +30,83 @@
...
+	PWR_CR3 = (PWR_CR3 & ~cr3_mask) | (PWR_CR3_SCUEN | PWR_CR3_LDOEN);
+}
+
+void pwr_set_mode_smps_ldo(bool supply_external, uint32_t smps_level, bool use_ldo) {
+	uint32_t cr3_mask, cr3_set;
+	cr3_mask = (PWR_CR3_SMPSEXTHP | PWR_CR3_SMPSEN | PWR_CR3_LDOEN | PWR_CR3_BYPASS);
+	cr3_mask |= PWR_CR3_SMPSLEVEL_MASK << PWR_CR3_SMPSLEVEL_SHIFT;
+
+	/* Default, take in unconditional settings, will OR in the rest. */
+	cr3_set = PWR_CR3_SMPSEN | (smps_level << PWR_CR3_SMPSLEVEL_SHIFT);
```
Comment (karlp) [inline lib/stm32/h7/pwr.c:60]: you shift it here, (which is good, we like letting uses pass in unshifted values).....

File: lib/stm32/h7/pwr.c [hunk 7]
```diff
@@ -30,15 +30,83 @@
...
+			break;
+		case PWR_SYS_SCU_BYPASS:
+			pwr_set_mode_scu_bypass();
+			break;
+		case PWR_SYS_LDO:
+			pwr_set_mode_ldo();
+			break;
+		case PWR_SYS_SMPS_DIRECT:
+		case PWR_SYS_SMPS_LDO:
+			reg_level = PWR_CR3_SMPSLEVEL_VOS << PWR_CR3_SMPSLEVEL_SHIFT;
```
Comment (karlp) [inline lib/stm32/h7/pwr.c:97]: and you've shifted it again here.  (and again on line 102)  Can you not just drop the intermediary variable?

**Comment by karlp** [comment]
Sorry, I missed the shift earlier, everythign else looks good here now, thanks

File: lib/stm32/h7/pwr.c (hunk 7 above)
Comment (vielster) [reply lib/stm32/h7/pwr.c:97]: Nice catch! Will remove the redundant shift bug.

File: include/libopencm3/stm32/h7/pwr.h [hunk 8]
```diff
@@ -98,27 +196,49 @@ enum pwr_svos_scale {
...
+/** Set power subsystem to utilize the SMPS run through the LDO for CPU.
+ *  @param[in] supply_external  Supply is powering external circuits, enable high power mode.
+ *  @param[in] smps_level  Voltage level from available PWR_CR3_SMPSLEVEL_XXX settings (1.8V/2.5V)
+ *  @param[in] use_ldo  Set this value to true if the internal LDO should be enabled.
+ */
+void pwr_set_mode_smps_ldo(bool supply_external, uint32_t smps_level, bool use_ldo);
+
+/** Set power system based on "System Supply Configurations" table in reference manual.
+ *  @param[in] mode  Mode mapping to a mode in the system configuration. Note special SCU modes.
+ *  @param[in] smps_level  Optional, only needed if using an EXT_SMPS or SMPS_LDO mode.
```
Comment (karlp) [inline include/libopencm3/stm32/h7/pwr.h:226]: What is "optional" here, are you mean to provide zero?  is this from PWR_CR3_SMPSLEVEL_xxxx as well? can I add doc groups for them?  (Don't worry, I'm merging this, it looks excelltne thanks, just want to tidy up the docs as much as I can while it's fresh)

Comment (karlp) [reply include/libopencm3/stm32/h7/pwr.h:226]: See https://github.com/libopencm3/libopencm3/pull/1307 for how I think I've understood this.

**Comment by karlp** [comment]
this has merged in c2dbea012bb8 thanks!

**Comment by donotnoot** [comment]
@vielster and @karlp this works perfectly well on a STM32H735 too :) thanks!
